File: src/main/java/no/nav/sosialhjelp/soknad/business/service/soknadservice/InnsendtSoknadService.java
package no.nav.sosialhjelp.soknad.business.service.soknadservice;

import no.nav.sosialhjelp.soknad.business.domain.BehandlingsKjede;
import no.nav.sosialhjelp.soknad.business.domain.BehandlingsKjede.InnsendtSoknad;
import no.nav.sosialhjelp.soknad.business.domain.SoknadMetadata;
import no.nav.sosialhjelp.soknad.business.domain.SoknadMetadata.VedleggMetadata;
import no.nav.sosialhjelp.soknad.business.service.HenvendelseService;
import no.nav.sosialhjelp.soknad.domain.SoknadMetadataInnsendingStatus;
import no.nav.sosialhjelp.soknad.domain.Vedleggstatus;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Predicate;

import static java.util.stream.Collectors.toList;
import static no.nav.sosialhjelp.soknad.business.util.EttersendelseUtils.soknadSendtForMindreEnn30DagerSiden;
import static no.nav.sosialhjelp.soknad.domain.model.kravdialoginformasjon.SoknadType.SEND_SOKNAD_KOMMUNAL_ETTERSENDING;

@Component
public class InnsendtSoknadService {

    public static final String SKJEMANUMMER_KVITTERING = "L7";

    private final HenvendelseService henvendelseService;

    private Predicate<VedleggMetadata> ikkeKvittering = v -> !SKJEMANUMMER_KVITTERING.equals(v.skjema);
    private Predicate<VedleggMetadata> lastetOpp = v -> v.status.er(Vedleggstatus.LastetOpp);
    private Predicate<VedleggMetadata> ikkeLastetOpp = lastetOpp.negate();

    private DateTimeFormatter datoFormatter = DateTimeFormatter.ofPattern("d. MMMM yyyy");
    private DateTimeFormatter tidFormatter = DateTimeFormatter.ofPattern("HH:mm:ss");

    public InnsendtSoknadService(HenvendelseService henvendelseService) {
        this.henvendelseService = henvendelseService;
    }

    static long soknadsalderIMinutter(LocalDateTime tidspunktSendt) {
        if (tidspunktSendt == null) return -1;
        return tidspunktSendt.until(LocalDateTime.now(), ChronoUnit.MINUTES);
    }

    public BehandlingsKjede hentBehandlingskjede(String behandlingsId) {
        SoknadMetadata originalSoknad = hentOriginalSoknad(behandlingsId);
        List<SoknadMetadata> ettersendelser = hentEttersendelser(originalSoknad.behandlingsId);

        return new BehandlingsKjede()
                .medOriginalSoknad(konverter(originalSoknad))
                .medEttersendelser(ettersendelser.stream()
                        .map(this::konverter)
                        .collect(toList())
                );
    }

    private SoknadMetadata hentOriginalSoknad(String behandlingsId) {
        SoknadMetadata soknad = henvendelseService.hentSoknad(behandlingsId);
        if (soknad.type == SEND_SOKNAD_KOMMUNAL_ETTERSENDING) {
            soknad = henvendelseService.hentSoknad(soknad.tilknyttetBehandlingsId);
        }
        return soknad;
    }

    private List<SoknadMetadata> hentEttersendelser(String behandlingsId) {
        return henvendelseService.hentBehandlingskjede(behandlingsId).stream()
                .filter(soknad -> soknad.status.equals(SoknadMetadataInnsendingStatus.FERDIG))
                .sorted(Comparator.comparing(o -> o.innsendtDato))
                .collect(toList());
    }

    private InnsendtSoknad konverter(SoknadMetadata metadata) {
        return new InnsendtSoknad()
                .medBehandlingId(metadata.behandlingsId)
                .medInnsendtDato(metadata.innsendtDato.format(datoFormatter))
                .medInnsendtTidspunkt(metadata.innsendtDato.format(tidFormatter))
                .medSoknadsalderIMinutter(soknadsalderIMinutter(metadata.innsendtDato))
                .medNavenhet(metadata.navEnhet)
                .medOrgnummer(metadata.orgnr)
                .medInnsendteVedlegg(tilVedlegg(metadata.vedlegg.vedleggListe, lastetOpp))
                .medIkkeInnsendteVedlegg(soknadSendtForMindreEnn30DagerSiden(metadata.innsendtDato.toLocalDate()) ? tilVedlegg(metadata.vedlegg.vedleggListe, ikkeLastetOpp) : null);
    }

    public LocalDateTime getInnsendingstidspunkt(String behandlingsId) {
        var soknadMetadata = hentOriginalSoknad(behandlingsId);
        return soknadMetadata.innsendtDato;
    }

    private List<InnsendtSoknad.Vedlegg> tilVedlegg(List<VedleggMetadata> vedlegg, Predicate<VedleggMetadata> status) {
        List<VedleggMetadata> vedleggMedRiktigStatus = vedlegg.stream()
                .filter(ikkeKvittering)
                .filter(status)
                .collect(toList());

        Map<String, InnsendtSoknad.Vedlegg> unikeVedlegg = new HashMap<>();

        vedleggMedRiktigStatus.forEach(v -> {
            String sammensattnavn = v.skjema + "|" + v.tillegg;
            if (!unikeVedlegg.containsKey(sammensattnavn)) {
                unikeVedlegg.put(sammensattnavn, new InnsendtSoknad.Vedlegg(
                        v.skjema,
                        v.tillegg,
                        v.status
                ));
            }
        });

        return new ArrayList<>(unikeVedlegg.values());
    }

}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/ettersending/innsendtsoknad/InnsendtSoknadService.kt
package no.nav.sosialhjelp.soknad.ettersending.innsendtsoknad

import no.nav.sosialhjelp.soknad.ettersending.innsendtsoknad.EttersendelseUtils.soknadSendtForMindreEnn30DagerSiden
import no.nav.sosialhjelp.soknad.repository.soknadmetadata.SoknadMetadata
import no.nav.sosialhjelp.soknad.repository.soknadmetadata.SoknadMetadataInnsendingStatus
import no.nav.sosialhjelp.soknad.repository.soknadmetadata.SoknadMetadataRepository
import no.nav.sosialhjelp.soknad.repository.soknadmetadata.SoknadMetadataType
import no.nav.sosialhjelp.soknad.repository.soknadmetadata.VedleggMetadata
import no.nav.sosialhjelp.soknad.repository.soknadmetadata.Vedleggstatus
import org.springframework.stereotype.Component
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import java.util.function.Predicate

@Component
class InnsendtSoknadService(
    private val soknadMetadataRepository: SoknadMetadataRepository
) {
    private val ikkeKvittering = Predicate<VedleggMetadata> { SKJEMANUMMER_KVITTERING != it.skjema }
    private val lastetOpp = Predicate<VedleggMetadata> { it.status?.er(Vedleggstatus.LastetOpp) ?: false }
    private val ikkeLastetOpp = lastetOpp.negate()
    private val datoFormatter = DateTimeFormatter.ofPattern("d. MMMM yyyy")
    private val tidFormatter = DateTimeFormatter.ofPattern("HH:mm:ss")

    fun hentBehandlingskjede(behandlingsId: String): BehandlingsKjede {
        val originalSoknad = hentOriginalSoknad(behandlingsId)
        val ettersendelser = originalSoknad?.let { hentEttersendelser(it.behandlingsId) }
        return BehandlingsKjede(
            originalSoknad = originalSoknad?.let { konverter(it) },
            ettersendelser = ettersendelser?.map { konverter(it) }
        )
    }

    private fun hentOriginalSoknad(behandlingsId: String): SoknadMetadata? {
        var soknad = soknadMetadataRepository.hent(behandlingsId)
        if (soknad?.type == SoknadMetadataType.SEND_SOKNAD_KOMMUNAL_ETTERSENDING) {
            soknad = soknadMetadataRepository.hent(soknad.tilknyttetBehandlingsId)
        }
        return soknad
    }

    private fun hentEttersendelser(behandlingsId: String): List<SoknadMetadata> {
        return soknadMetadataRepository.hentBehandlingskjede(behandlingsId)
            .filter { it.status == SoknadMetadataInnsendingStatus.FERDIG }
            .sortedWith(Comparator.comparing { it.innsendtDato })
    }

    private fun konverter(metadata: SoknadMetadata): InnsendtSoknad {
        return InnsendtSoknad(
            behandlingsId = metadata.behandlingsId,
            innsendtDato = metadata.innsendtDato?.format(datoFormatter),
            innsendtTidspunkt = metadata.innsendtDato?.format(tidFormatter),
            soknadsalderIMinutter = soknadsalderIMinutter(metadata.innsendtDato),
            innsendteVedlegg = metadata.vedlegg?.vedleggListe?.let { tilVedlegg(it, lastetOpp) },
            ikkeInnsendteVedlegg = metadata.innsendtDato?.toLocalDate()
                ?.let { if (soknadSendtForMindreEnn30DagerSiden(it)) tilVedlegg(metadata.vedlegg?.vedleggListe, ikkeLastetOpp) else null },
            navenhet = metadata.navEnhet,
            orgnummer = metadata.orgnr
        )
    }

    fun getInnsendingstidspunkt(behandlingsId: String): LocalDateTime? {
        val soknadMetadata = hentOriginalSoknad(behandlingsId)
        return soknadMetadata?.innsendtDato
    }

    private fun tilVedlegg(
        vedlegg: List<VedleggMetadata>?,
        status: Predicate<VedleggMetadata>
    ): List<Vedlegg> {
        val vedleggMedRiktigStatus = vedlegg
            ?.filter { ikkeKvittering.test(it) }
            ?.filter { status.test(it) }

        val unikeVedlegg: MutableMap<String, Vedlegg> = HashMap()
        vedleggMedRiktigStatus?.forEach {
            val sammensattnavn = it.skjema + "|" + it.tillegg
            if (!unikeVedlegg.containsKey(sammensattnavn)) {
                unikeVedlegg[sammensattnavn] = Vedlegg(
                    skjemaNummer = it.skjema,
                    skjemanummerTillegg = it.tillegg,
                    innsendingsvalg = it.status
                )
            }
        }
        return unikeVedlegg.values.toList()
    }

    companion object {
        const val SKJEMANUMMER_KVITTERING = "L7"

        fun soknadsalderIMinutter(tidspunktSendt: LocalDateTime?): Long {
            return tidspunktSendt?.until(LocalDateTime.now(), ChronoUnit.MINUTES) ?: -1
        }
    }
}
